### PR TITLE
Remove dependency between build-cache-packaging and native

### PR DESCRIPTION
**Pull request by britter**
By replacing the usage of FileSystem with a little utility class we can
get rid of the dependency between build-cache-packaging and native. For
this to work the TarBuildCacheEntryPackerPermissionTest hat to be
changes to assert correct permissions instead of checking for
interactions.

Part of https://github.com/gradle/gradle-private/issues/2417

### CODE REVIEW

File: subprojects/build-cache-packaging/src/main/java/org/gradle/caching/internal/packaging/impl/TarBuildCacheEntryPacker.java [hunk 1]
```diff
@@ -322,7 +319,11 @@ private TarArchiveEntry unpackDirectoryTree(TarArchiveInputStream input, TarArch
     }
 
     private void chmodUnpackedFile(TarArchiveEntry entry, File file) {
-        fileSystem.chmod(file, entry.getMode() & UnixPermissions.PERM_MASK);
+        try {
+            PermissionUtils.setPermissions(file, entry.getMode() & UnixPermissions.PERM_MASK);
+        } catch (IOException e) {
+            throw new java.io.UncheckedIOException(e);
```
Comment (lptr) [inline subprojects/build-cache-packaging/src/main/java/org/gradle/caching/internal/packaging/impl/TarBuildCacheEntryPacker.java]: Import.

File: subprojects/core/core.gradle.kts [hunk 2]
```diff
@@ -100,6 +100,9 @@ dependencies {
     testFixturesApi(project(":execution")) {
         because("test fixtures expose OutputChangeListener")
     }
+    testFixturesApi(project(":native")) {
+        because("test fixtures expose FileSystem")
+    }
```
Comment (lptr) [inline subprojects/core/core.gradle.kts]: Why is this needed now?

File: subprojects/build-cache-packaging/src/test/groovy/org/gradle/caching/internal/packaging/impl/TarBuildCacheEntryPackerPermissionTest.groovy [hunk 3]
```diff
@@ -35,29 +30,25 @@ class TarBuildCacheEntryPackerPermissionTest extends AbstractTarBuildCacheEntryP
...
+        and:
         def input = new ByteArrayInputStream(output.toByteArray())
         unpack input, prop(FILE, targetOutputFile)
 
         then:
         targetOutputFile.text == "output"
-        1 * fileSystem.chmod(targetOutputFile, unixMode)
-        _ * targetOutputFile._
-        0 * _
+        PermissionUtils.permissions(targetOutputFile) == unixMode
```
Comment (lptr) [inline subprojects/build-cache-packaging/src/test/groovy/org/gradle/caching/internal/packaging/impl/TarBuildCacheEntryPackerPermissionTest.groovy]: Will these tests pass on Windows as well?

File: subprojects/build-cache-packaging/src/main/java/org/gradle/caching/internal/packaging/impl/PermissionUtils.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+        posixPermissionToInteger.put(PosixFilePermission.GROUP_READ, 0040);
+
+        posixPermissionToInteger.put(PosixFilePermission.OTHERS_EXECUTE, 0001);
+        posixPermissionToInteger.put(PosixFilePermission.OTHERS_WRITE, 0002);
+        posixPermissionToInteger.put(PosixFilePermission.OTHERS_READ, 0004);
+    }
+
+    public static void setPermissions(File file, int mode) throws IOException {
+        if (IS_POSIX) {
+            Files.setPosixFilePermissions(file.toPath(), PermissionUtils.permissions(mode));
```
Comment (lptr) [inline subprojects/build-cache-packaging/src/main/java/org/gradle/caching/internal/packaging/impl/PermissionUtils.java]: Is it necessary to qualify this call with `PermissionUtils.`?

File: subprojects/core/core.gradle.kts (hunk 2 above)
Comment (britter) [reply subprojects/core/core.gradle.kts]: Without this I was getting, I suspect because of AST transformations or something like that.

```
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':buildCachePackaging:compileTestGroovy'.
Caused by: java.lang.NoClassDefFoundError: org.gradle.internal.nativeintegration.filesystem.FileSystem
```

This declaration was missing in core because of `org.gradle.api.internal.file.TestFiles#fileSystem()` which returns a `FileSystem`

File: subprojects/build-cache-packaging/src/test/groovy/org/gradle/caching/internal/packaging/impl/TarBuildCacheEntryPackerPermissionTest.groovy (hunk 3 above)
Comment (britter) [reply subprojects/build-cache-packaging/src/test/groovy/org/gradle/caching/internal/packaging/impl/TarBuildCacheEntryPackerPermissionTest.groovy]: If you're asking like this I suspect not 😇 . Should we ignore it on Windows?

Comment (lptr) [reply subprojects/build-cache-packaging/src/test/groovy/org/gradle/caching/internal/packaging/impl/TarBuildCacheEntryPackerPermissionTest.groovy]: Perhaps a better option would be to turn `PermissionUtils` into a service, so we could mock it here. Then we could add a separate test for Windows and Linux/macOS for this `PermissionSetter` service.

File: subprojects/core/core.gradle.kts (hunk 2 above)
Comment (lptr) [reply subprojects/core/core.gradle.kts]: Got it, thanks.

File: subprojects/build-cache-packaging/src/test/groovy/org/gradle/caching/internal/packaging/impl/TarBuildCacheEntryPackerPermissionTest.groovy (hunk 3 above)
Comment (britter) [reply subprojects/build-cache-packaging/src/test/groovy/org/gradle/caching/internal/packaging/impl/TarBuildCacheEntryPackerPermissionTest.groovy]: Looking at the code again, it does something with `DosFilePermissions`. So maybe it even works on Windows...

File: subprojects/build-cache-packaging/src/main/java/org/gradle/caching/internal/packaging/impl/TarBuildCacheEntryPacker.java (hunk 1 above)
Comment (britter) [reply subprojects/build-cache-packaging/src/main/java/org/gradle/caching/internal/packaging/impl/TarBuildCacheEntryPacker.java]: Done

File: subprojects/build-cache-packaging/src/main/java/org/gradle/caching/internal/packaging/impl/PermissionUtils.java (hunk 4 above)
Comment (britter) [reply subprojects/build-cache-packaging/src/main/java/org/gradle/caching/internal/packaging/impl/PermissionUtils.java]: Removed it

File: subprojects/build-cache-packaging/src/test/groovy/org/gradle/caching/internal/packaging/impl/TarBuildCacheEntryPackerPermissionTest.groovy (hunk 3 above)
Comment (britter) [reply subprojects/build-cache-packaging/src/test/groovy/org/gradle/caching/internal/packaging/impl/TarBuildCacheEntryPackerPermissionTest.groovy]: @lptr it passed the ready for merge pipeline so I guess it works on windows? If it wouldn't work on Windows I don't see much value in the additional abstraction to make something testable on Windows that does not work with the real implementation.

The advantage of having an additional interface would be that we could continue using `FileSystem` (via an adapter) instead of having the some functionality implemented in two places.

**Comment by wolfs** [comment]
@britter Note that the published artifacts from Gradle now have transitive dependencies, so you might want to use those instead of declaring all the dependencies one by one.

File: subprojects/build-cache-packaging/src/test/groovy/org/gradle/caching/internal/packaging/impl/TarBuildCacheEntryPackerPermissionTest.groovy (hunk 3 above)
Comment (lptr) [reply subprojects/build-cache-packaging/src/test/groovy/org/gradle/caching/internal/packaging/impl/TarBuildCacheEntryPackerPermissionTest.groovy]: I don't see any `Ready for Merge` builds executed for this branch, only the performance regression tests, and those don't involve Windows tests. I fired one off now: https://builds.gradle.org/viewQueued.html?itemId=30121516

If that build passes I'm okay with merging this as-is, and continue refining the API for `PermissionUtils` in a followup PR.

Comment (lptr) [reply subprojects/build-cache-packaging/src/test/groovy/org/gradle/caching/internal/packaging/impl/TarBuildCacheEntryPackerPermissionTest.groovy]: Yep, fails on Windows: https://builds.gradle.org/viewLog.html?buildId=30121365

Comment (lptr) [reply subprojects/build-cache-packaging/src/test/groovy/org/gradle/caching/internal/packaging/impl/TarBuildCacheEntryPackerPermissionTest.groovy]: We can do what you suggested, and put `FileSystem` behind a wrapper interface for Gradle, and use `PermissionUtils` in Maven.

Though that said I'd love to have a single implementation for the service using JDK 7 APIs, and some comprehensive tests to demonstrate what happens to permissions on Windows, Linux and macOS each. Maybe not something we need to do for this PR, though.

Comment (britter) [reply subprojects/build-cache-packaging/src/test/groovy/org/gradle/caching/internal/packaging/impl/TarBuildCacheEntryPackerPermissionTest.groovy]: Okay, so should we close this PR and create a new one introducing an interface?

I still don't get what you like to test on windows. As far as I understand the code does nothing on windows, no?

Comment (lptr) [reply subprojects/build-cache-packaging/src/test/groovy/org/gradle/caching/internal/packaging/impl/TarBuildCacheEntryPackerPermissionTest.groovy]: I'm not 100% sure right now. Let's aim for keeping stuff as it is, and extract the interface, and use what we currently use underneath on both sides for now. That'll unblock us, and we can always come back to improve the situation if we think we need to.

Comment (britter) [reply subprojects/build-cache-packaging/src/test/groovy/org/gradle/caching/internal/packaging/impl/TarBuildCacheEntryPackerPermissionTest.groovy]: I will create a new PR for this.

**Comment by britter** [comment]
As discussed above we will go with a different approach.
